test/stop/residual_norm_kernels.cpp

```diff
@@ -0,0 +1,783 @@
+DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY
+THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
+(INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
+OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+******************************<GINKGO LICENSE>*******************************/
+
+#include <gtest/gtest.h>
+
+
+#include <ginkgo/core/base/math.hpp>
+#include <ginkgo/core/stop/residual_norm.hpp>
+
+
+#include "core/test/utils.hpp"
+#include "test/utils/executor.hpp"
+
+
+template <typename Mtx>
+void write(std::unique_ptr<Mtx>& mtx, gko::size_type row, gko::size_type col,
+           typename Mtx::value_type val)
+{
+    mtx->get_executor()->copy_from(
+        mtx->get_executor()->get_master().get(), 1, &val,
+        mtx->get_values() + col + mtx->get_stride() * row);
+}
+
+
+template <typename Mtx>
+typename Mtx::value_type read(std::unique_ptr<Mtx>& mtx, gko::size_type row,
+                              gko::size_type col)
+{
+    return mtx->get_executor()->copy_val_to_host(mtx->get_values() + col +
+                                                 mtx->get_stride() * row);
+}
+
+
+template <typename T>
+class ResidualNorm : public CommonTestFixture {
+protected:
+    using Mtx = gko::matrix::Dense<T>;
+    using NormVector = gko::matrix::Dense<gko::remove_complex<T>>;
+
+    ResidualNorm()
+    {
+        factory = gko::stop::ResidualNorm<T>::build()
+                      .with_reduction_factor(r<T>::value)
+                      .on(exec);
+        rel_factory = gko::stop::ResidualNorm<T>::build()
+                          .with_reduction_factor(r<T>::value)
+                          .with_baseline(gko::stop::mode::initial_resnorm)
+                          .on(exec);
+        abs_factory = gko::stop::ResidualNorm<T>::build()
+                          .with_reduction_factor(r<T>::value)
+                          .with_baseline(gko::stop::mode::absolute)
+                          .on(exec);
+    }
+
+    std::unique_ptr<typename gko::stop::ResidualNorm<T>::Factory> factory;
+    std::unique_ptr<typename gko::stop::ResidualNorm<T>::Factory> rel_factory;
+    std::unique_ptr<typename gko::stop::ResidualNorm<T>::Factory> abs_factory;
+};
+
+TYPED_TEST_SUITE(ResidualNorm, gko::test::ValueTypes, TypenameNameGenerator);
+
+
+TYPED_TEST(ResidualNorm, WaitsTillResidualGoal)
+{
+    using Mtx = typename TestFixture::Mtx;
+    using NormVector = typename TestFixture::NormVector;
+    auto initial_res = gko::initialize<Mtx>({100.0}, this->exec);
+    std::shared_ptr<gko::LinOp> rhs = gko::initialize<Mtx>({10.0}, this->exec);
+    auto criterion =
+        this->factory->generate(nullptr, rhs, nullptr, initial_res.get());
+    auto rel_criterion =
+        this->rel_factory->generate(nullptr, rhs, nullptr, initial_res.get());
+    auto abs_criterion =
+        this->abs_factory->generate(nullptr, rhs, nullptr, initial_res.get());
+    {
```

Comment:
the following three scope does not share too much common factor, maybe separate them to three test?